genoray/_svar.py

```diff
@@ -1067,3 +1188,187 @@ def _find_starts_ends_with_length(
     out = out[:, unsorter]
 
     return out
+
+
+def _empty_annot() -> pl.DataFrame:
+    """Return an empty annotation DataFrame with the correct schema."""
+    return pl.DataFrame(
+        {"varID": [], "gene_id": [], "strand": [], "codon_pos": []},
+        schema={
+            "varID": pl.UInt32,
+            "gene_id": pl.Utf8,
+            "strand": pl.Utf8,
+            "codon_pos": pl.Int8,
+        },
+    )
+
+
+def _get_strand_and_codon_pos(
+    cds_df: pl.DataFrame,
+    granges: pr.PyRanges,
+    ilen_df: pl.DataFrame,
+    contig_normalizer: ContigNormalizer,
+) -> pl.DataFrame:
+    """
+    Calculate strand and codon position for variants overlapping CDS regions.
+
+    Parameters
+    ----------
+    cds_df : pl.DataFrame
+        CDS features from GTF with columns: chrom, start, end, strand, frame,
+        gene_id, transcript_id, gene_biotype, transcript_support_level, tag
+    granges : pr.PyRanges
+        Variant genomic ranges
+    ilen_df : pl.DataFrame
+        DataFrame with varID and ILEN columns
+    contig_normalizer : ContigNormalizer
+        Normalizer to match chromosome names between CDS and granges
+
+
+    Returns
+    -------
+    pl.DataFrame
+        Annotation with varID, gene_id, strand, codon_pos
+    """
+
+    # Normalize CDS chromosome names to match granges
+    # Cast to string first to avoid categorical comparison issues
+    cds_df = cds_df.with_columns(pl.col("chrom").cast(pl.Utf8))
+
+    chrom_list = cds_df["chrom"].unique().to_list()
+    normalized = contig_normalizer.norm(chrom_list)
+    norm_map = {orig: norm or orig for orig, norm in zip(chrom_list, normalized)}
+
+    cds_df = cds_df.with_columns(
+        pl.col("chrom").replace_strict(norm_map, default=pl.col("chrom"))
+    )
+
+    # Filter out CDS features with chromosomes not in granges
+    cds_df = cds_df.filter(pl.col("chrom").is_in(contig_normalizer.contigs))
+
+    pr_cds = _to_pyranges(cds_df)
+    joined_cds = granges.join(pr_cds, apply_strand_suffix=False)
+
+    if len(joined_cds) == 0:
+        return _empty_annot()
+
+    annot = (
+        pl.from_pandas(joined_cds.df)
+        .with_columns(
+            pl.col("index").cast(pl.UInt32).alias("varID"),
+            # Keep PyRanges 0-based coordinates for direct offset calculation
+            pl.col("Start").alias("pos"),
+            pl.col("Start_b").alias("cds_start"),
+            # Convert PyRanges 0-based exclusive End_b to 0-based inclusive cds_end
+            (pl.col("End_b") - 1).alias("cds_end"),
```

Comment:
```suggestion
            cds_end="End_b",
```

cds end is just used to compute feature length for breaking rank ties later

Reply:
makes sense, updated this